testUpdateHostsFile.py

```diff
@@ -1616,44 +1547,43 @@ def test_line_without_prefix(self):
 
 
 class GetFileByUrl(BaseStdout):
-    @mock.patch("updateHostsFile.urlopen", side_effect=mock_url_open)
-    def test_read_url(self, _):
-        url = b"www.google.com"
-
-        expected = "www.google.com"
-        actual = get_file_by_url(url, delay=0)
-
-        self.assertEqual(actual, expected)
+    def test_basic(self):
+        raw_resp_content = "hello, ".encode("ascii") + "world".encode("utf-8")
+        resp_obj = requests.Response()
+        resp_obj.__setstate__({"_content": raw_resp_content})
 
-    @mock.patch("updateHostsFile.urlopen", side_effect=mock_url_open_fail)
-    def test_read_url_fail(self, _):
-        url = b"www.google.com"
-        self.assertIsNone(get_file_by_url(url, delay=0))
+        expected = "hello, world"
 
-        expected = "Problem getting file:"
-        output = sys.stdout.getvalue()
+        with mock.patch("requests.get", return_value=resp_obj):
+            actual = get_file_by_url("www.test-url.com")
 
-        self.assertIn(expected, output)
+        self.assertEqual(expected, actual)
 
-    @mock.patch("updateHostsFile.urlopen", side_effect=mock_url_open_read_fail)
-    def test_read_url_read_fail(self, _):
-        url = b"www.google.com"
-        self.assertIsNone(get_file_by_url(url, delay=0))
+    def test_with_idna(self):
+        raw_resp_content = b"www.huala\xc3\xb1e.cl"
+        resp_obj = requests.Response()
+        resp_obj.__setstate__({"_content": raw_resp_content})
 
-        expected = "Problem getting file:"
-        output = sys.stdout.getvalue()
+        expected = "www.xn--hualae-0wa.cl"
 
-        self.assertIn(expected, output)
+        with mock.patch("requests.get", return_value=resp_obj):
+            actual = get_file_by_url("www.test-url.com")
 
-    @mock.patch("updateHostsFile.urlopen", side_effect=mock_url_open_decode_fail)
-    def test_read_url_decode_fail(self, _):
-        url = b"www.google.com"
-        self.assertIsNone(get_file_by_url(url, delay=0))
+        self.assertEqual(expected, actual)
 
-        expected = "Problem getting file:"
-        output = sys.stdout.getvalue()
+    def test_connect_unknown_domain(self):
+        test_url = "http://doesnotexist.google.com"
+        return_value = get_file_by_url(test_url)
+        self.assertIsNone(return_value)
+        printed_output = sys.stdout.getvalue()
+        self.assertEqual(printed_output, "Error retrieving data from {}\n".format(test_url))
 
-        self.assertIn(expected, output)
+    def test_invalid_url(self):
+        test_url = "http://fe80::5054:ff:fe5a:fc0"
+        return_value = get_file_by_url(test_url)
+        self.assertIsNone(return_value)
+        printed_output = sys.stdout.getvalue()
+        self.assertEqual(printed_output, "Error retrieving data from {}\n".format(test_url))
```

Comment:
Can we mock the responses instead of using live calls to `request.get` ?

Reply:
That's probably a good idea, yes. I'll dig through the code for Requests to see if there's a convenient place we can mock.

Reply:
I'm struggling to find the best place to patch/mock, do you have any ideas? If I go too low, I increase the chances that the tests will break on a different OS, Python version, etc. It seems pointless to go much higher, though, since the code is trivial anyway, and if something fails there then we have much bigger issues to worry about.

Reply:
Is it possible to mock the `Response` object that you get back from `requests.get` ?

Reply:
Yes, it's what I'm doing in one of the test cases. The issue is with the ones which are supposed to lead to an exception.

Reply:
You can mock the `requests.get` call to raise an exception.

Reply:
Is it even worth it at that point? All we're testing is whether or not the try-except catches the exception.

Reply:
IMO yes.  `try-except` is still functionality that should be tested.

Reply:
Done! :)